### PR TITLE
lightning: tidb backend will check conflict threshold

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/41629

Problem Summary:

### What is changed and how it works?

use affected rows to check if batch insert meets duplications

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by dsdashun — APPROVED** [review]

File: br/pkg/lightning/backend/tidb/tidb.go [hunk 1]
```diff
@@ -276,12 +276,29 @@ var _ backend.Backend = (*tidbBackend)(nil)
...
+		onDuplicate = config.ErrorOnDup
+	case config.ReplaceOnDup:
 		onDuplicate = config.ReplaceOnDup
+	case config.IgnoreOnDup:
+		if conflict.MaxRecordRows == 0 {
+			onDuplicate = config.IgnoreOnDup
+		} else {
+			// need to stop batch insert on error and fall back to row by row insert
+			// to record the row
+			onDuplicate = config.ErrorOnDup
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/tidb/tidb.go:297]:
```
mysql> insert into t values(1);
ERROR 1062 (23000): Duplicate entry '1' for key 't.PRIMARY'
mysql> insert ignore into t values(1);
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql>
mysql> replace into t values(1);
Query OK, 1 row affected (0.00 sec)
```
`insert ingore into` does have different RowAffected as len(rows), but `replace into` not, shouldn't we use this checking on `replace`?

Comment (lance6716) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: > The REPLACE statement returns a count to indicate the number of rows affected. This is the sum of the rows deleted and inserted. If the count is 1 for a single-row REPLACE, a row was inserted and no rows were deleted. If the count is greater than 1, one or more old rows were deleted before the new row was inserted. It is possible for a single row to replace more than one old row if the table contains multiple unique indexes and the new row duplicates values for different old rows in different unique indexes.

The affected-rows count makes it easy to determine whether REPLACE only added a row or whether it also replaced any rows: Check whether the count is 1 (added) or greater (replaced).

https://dev.mysql.com/doc/refman/8.4/en/replace.html

seems the behavior is not the same as documents

Comment (D3Hunter) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: I tried a real replaced case, it works. and it seems we can also tell how many rows ignored using RowAffected
```
mysql> create table t(id int primary key, v int);
Query OK, 0 rows affected (0.15 sec)

mysql> insert into t values(1,1);
Query OK, 1 row affected (0.00 sec)

mysql> replace into t values(1,1);
Query OK, 1 row affected (0.00 sec)

mysql> replace into t values(2,2);
Query OK, 1 row affected (0.00 sec)

mysql> replace into t values(1,2);
Query OK, 2 rows affected (0.00 sec)

mysql> insert ignore into t values(3,3);
Query OK, 1 row affected (0.00 sec)

mysql> insert ignore into t values(1,1);
Query OK, 0 rows affected, 1 warning (0.00 sec)
```

Comment (lance6716) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: I should have implemented it. When the user doesn't require conflict row details, lightning will use REPLACE and check affected rows to calculate if conflict threshold is exceeded

Comment (D3Hunter) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: yes, but for `ignore` we will set it to `error` if `conflict.threshold > 0`, it seems we can implement the same as `replace`

Comment (lance6716) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: we will use `IgnoreOnDup` when `conflict.MaxRecordRows == 0` (line 292). But when `conflict.MaxRecordRows` is not 0, we need to know which row is conflicting to record it, so we can't use batched INSERT IGNORE INTO.

Comment (D3Hunter) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: why not just record a count like `replace`, for compatibility?

Comment (lance6716) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: This is two features. One is set an early fail threshold for large amount of duplicate data. The other one is display the duplicate data.

Comment (lance6716) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: Displaying the duplicate data has more details than only limit the count, so it causes worse performance (fallback to row-by-row INSERT) and feature compatibility (can't use REPLACE)

Comment (D3Hunter) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: get, but i think we can make `ignore` works like `replace` to record count only later. I met a case that user uses `ignore` but it still print `ERROR` logs on duplicate and the SQL is `INSERT INTO` unlike the doc says as `IGNORE INTO`(it's fixed here https://github.com/pingcap/docs/pull/18911/files), it's confusing though i get the point after checking the code

Comment (lance6716) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: We can achieve it if it's 0 MaxRecordRows. Maybe change the default value to 0? Currently it's 10000

```golang
	if c.Threshold < 0 {
		switch c.Strategy {
		case ErrorOnDup, NoneOnDup:
			c.Threshold = 0
		case IgnoreOnDup, ReplaceOnDup:
			c.Threshold = DefaultRecordDuplicateThreshold
		}
	}
```

Comment (D3Hunter) [reply br/pkg/lightning/backend/tidb/tidb.go:297]: lgtm

**Pull request merged**